tests-e2e/lists_test.go

```diff
@@ -856,3 +856,121 @@ func TestMultipleSameList(t *testing.T) {
+		t,
+		map[string]interface{}{
+			"preset": "public_chat",
+			"name":   "room2",
+		},
+	)
+	t.Logf("room1=%s room2=%s", room1, room2)
+	t.Log("Bob joins both rooms.")
+	bob.JoinRoom(t, room1, nil)
+	bob.JoinRoom(t, room2, nil)
+
+	t.Log("Bob sends a message in room 2 then room 1.")
+	bob.SendEventSynced(t, room2, Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"body":    "Hi room 2",
+			"msgtype": "m.text",
+		},
+	})
+	bob.SendEventSynced(t, room1, Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"body":    "Hello world",
+			"msgtype": "m.text",
+		},
+	})
+
+	t.Log("Alice requests a sliding sync that bumps rooms on messages only.")
+	aliceReqList := sync3.RequestList{
+		Sort:   []string{sync3.SortByRecency, sync3.SortByName},
+		Ranges: sync3.SliceRanges{{0, 20}},
+		RoomSubscription: sync3.RoomSubscription{
+			RequiredState: [][2]string{{"m.room.avatar", ""}, {"m.room.encryption", ""}},
+			TimelineLimit: 10,
+		},
+	}
+	aliceSyncRequest := sync3.Request{
+		Lists: map[string]sync3.RequestList{
+			"room_list": aliceReqList,
+		},
+		BumpEventTypes: []string{"m.room.message", "m.room.encrypted"},
+	}
+	// TODO: to avoid flakes, want this sync to include/cover Bob's messages in both rooms.
+	aliceRes := alice.SlidingSync(t, aliceSyncRequest)
+
+	t.Log("Alice's sync response should include room1 ahead of room 2.")
+	matchRoom1ThenRoom2 := m.MatchList("room_list",
+		m.MatchV3Count(2),
+		m.MatchV3Ops(
+			m.MatchV3SyncOp(0, 20, []string{room1, room2}, false),
+		))
+	m.MatchResponse(t, aliceRes, matchRoom1ThenRoom2)
+
+	t.Log("Bob requests a sliding sync that bumps rooms on messages and memberships.")
+	bobReqList := sync3.RequestList{
+		Sort:   []string{sync3.SortByRecency, sync3.SortByName},
+		Ranges: sync3.SliceRanges{{0, 20}},
+		RoomSubscription: sync3.RoomSubscription{
+			RequiredState: [][2]string{{"m.room.avatar", ""}, {"m.room.encryption", ""}},
+			TimelineLimit: 10,
+		},
+	}
+	bobSyncRequest := sync3.Request{
+		Lists: map[string]sync3.RequestList{
+			"room_list": bobReqList,
+		},
+		BumpEventTypes: []string{"m.room.message", "m.room.encrypted", "m.room.member"},
+	}
+	bobRes := bob.SlidingSync(t, bobSyncRequest)
+
+	t.Log("Bob should also see room 1 ahead of room 2 in his sliding sync response.")
+	m.MatchResponse(t, bobRes, matchRoom1ThenRoom2)
+
+	t.Log("Charlie joins room 2.")
+	charlie.JoinRoom(t, room2, nil)
+
+	t.Log("Alice syncs until she sees Charlie's membership.")
+	aliceRes = alice.SlidingSyncUntilMembership(t, aliceRes.Pos, room2, charlie, "join")
+
+	t.Log("Alice shouldn't see any rooms' positions change.")
+	matchNoOps := m.MatchList("room_list",
+		m.MatchV3Count(2),
+		m.MatchV3Ops(),
+	)
+	m.MatchResponse(t, aliceRes, matchNoOps)
+
+	t.Log("Bob syncs until he sees Charlie's membership.")
+	bobRes = bob.SlidingSyncUntilMembership(t, bobRes.Pos, room2, charlie, "join")
+
+	t.Log("Bob should see room 2 at the top of his list.")
```

Comment:
These logs are unhelpful when things fail because I at least ask "why?" and that isn't explained in the logs. Why should Bob see room 2 at the top of his list? Also, the dev can probably ascertain what is being asserted by just reading the code.

Reply:
I take the point that this doesn't spell out "why", but I think that can be inferred from looking at the full sequence of these signpost logs. (Which doesn't seem all that different to reading the source, to me.)

> These logs are unhelpful when things fail

Strong disagree here. We have found this pattern useful in complement, particularly in longer tests with lots of output. Pain points:

- not easy to see which clients are making CSAPI requests (e.g. who is syncing?)
- easy to miss a `t.Errorf` that happened midway through the test
- Errorf/Fatalf text isn't red or otherwise made more prominent (e.g. by being printed separately to the test logs)